generate enum type traits automatically.

hwanseung opened the pull request:
generate enum type traits automatically.

ISSUE=#80

romandev committed:
Initial commit for Bacardi project :)

romandev committed:
Add LICENSE.md file.

We are using Apache License Version 2.0.

romandev committed:
Import bootstrap codes from Absolute project.

From now, we can use |bacardi| command in three major platforms.

romandev committed:
Introduce build environment for C++ build.

This change includes the following things:
  - Use node-gyp to build Native codes.
  - Add examples directory to put .cc files.
  - Add some guides to build.

romandev committed:
Add .travis.xml for Linux build automation.

Introduce build automation for Linux platform.

romandev committed:
Add an calculator service example.

After this patch, we can build and test napi-modules simply.

romandev committed:
Introduce AUTHORS file.

The file has a list of all contributors for Bacardi.

romandev committed:
Add CODEOWNERS file.

Add CODEOWNERS file to define teams are responsible for code in a repo.

gogag2 committed:
Adding aguments to Add method.

Delete hard-coded value and pass aguments to process.

ISSUE=#4

romandev committed:
Support for Mac OSX.

Update README.md and .travis.yml in order to support for Mac OSX.

ISSUE=none

romandev committed:
Add clang-format tool for formatting C++ code to style.

After this patch, to automatically format a pending patch according to our
coding style, from the command line, simply run: ./bacardi format.

romandev committed:
Use bootstrap node command instead of global node command.

We should use bootstrap node command instead of global node command. If not so,
it might not work well. (e.g. not support n-api)

romandev committed:
Implement skeleton codes for generator.

After this patch, generator/generator.js can take idl files that defined in
binding.gyp.

gogag2 committed:
Test doesn't work after bacardi format.

The comparing value was wrong. So after formatting, test doesn't work.
We fixed the checking value.

ISSUE=none

hwanseung committed:
Use Assert module instead of console log. (#16)

this Assert module[1] is a default module of node.js.
so it does not need to additional install.

[1] https://nodejs.org/api/assert.html

ISSUE=none

gogag2 committed:
Adding sub, mul, div function for bacardi test. (#15)

We added sub, mul, div function to test bacardi project.
Then we will modularization them with next patch.

ISSUE=#10

gogag2 committed:
Upgrade nodejs from v8.2.1 to v8.4.0 (#20)

Upgrade nodejs from v8.2.1 to v8.4.0. In the latest version, napi_create_number
was deprecated. We should use napi_create_double instead[1]. Also,
napi_type_error's signature was changed[2];

[1] https://nodejs.org/api/n-api.html#n_api_napi_create_double
[2] https://nodejs.org/api/n-api.html#n_api_napi_throw_type_error

ISSUE=#18

romandev committed:
Introduce TypeScript for generator codes. (#21)

It's better to use TypeScript instead of JS with node.

ISSUE=#14

romandev committed:
Add WebIDL parser and simple code generator. (#23)

This change adds WebIDL parser and then implement a simple code generator. It
can not generate actual codes but it shows how code generator works.
(It will print a simple C++ class that maps with IDL file.)

ISSUE=#22

romandev committed:
Introduce Nunjucks template engine. (#24)

This change is introducing Nunjucks template engine that is essentially a port
of python jinja2. So, replaces existing simple output with rendered output by
Nunjucks with template file.

ISSUE=#22

nadongguri committed:
Fix templete not found error during build command (#25)

There's an error that cannot found example.tpl file during build.
I modified example.tpl to example.njk.

romandev committed:
Add include_dirs field in binding.gyp. (#28)

After this patch, we can use relative path from bacardi root directory.

ISSUE=none

romandev committed:
Implement initial codes for NativeTypeTraits. (#30)

This change is implementing an initial codes for NativeTypeTraits to convert
JS value to Native value easier. After this patch, we can use NativeTypeTraits
instead of individual converting in example codes.

ISSUE=#29

hwanseung committed:
Introduce macha which is test framework for node.js. (#27)

Mocha is a simple, flexible, fun JavaScript test framework for node.js and
the browser.

ISSUE=none

romandev committed:
Add ./bacardi clean command and use it. (#31)

Even although build is failed in travis bot, test is succeded sometimes.
So, we add a new clean command and then use it in travis script.

ISSUE=none

hwanseung committed:
Add package-lock.json to .gitignore (#32)

package-lock.json is automatically generated for any operations
where npm modifies either the node_modules tree, or package.json.
so sometimes it might be made to bother when use git command.

ISSUE=none

romandev committed:
Separate Init() function from calculator_bridge. (#33)

The Init() function is used to expose Native interfaces to JS layer. So, the
function should be separated from specific example codes(e.g. calculator).

ISSUE=none

romandev committed:
Move all methods into bacardi.Calculator namespace. (#34)

In the current calculator example, all static operations don't belong to any
class. It's a little bit weird because another example can be define in the
same scope. So, this change defines a new Calculator class in JS side and then
move all static methods into bacardi.Calculator namespace.

ISSUE=none

romandev committed:
Update README.md in more detail. (#35)

This change includes the following things:
  - Travis bot badge to display latest bot status for Linux and Mac.
  - How to test.
  - Coding style and how to format source codes automatically.

ISSUE=none

romandev committed:
Use node-addon-api instead of built-in N-API. (#36)

The package contains header-only C++ wrapper classes for the N-API, along with
library code that enables backward-compatibility with use with older versions
of Node.js that do not have N-API built-in.

ISSUE=none

romandev committed:
Migrate the calculator example to C++ code. (#40)

The calculator example is written in C++ language but it's also using built-in
N-API based on C language. On the other hand, we already introduced
node-addon-api in #36, so, we can just replace it with C++ wrapper code fully.

ISSUE=#37

romandev committed:
Make methods as instance methods in calculator example. (#41)

In the current calculator example, all methods are defined as static methods.
It's a little bit wierd because the class is not static only. So, this change
makes the methods as instance methods.

ISSUE=none

romandev committed:
Remove unnecessary TODO comments and fix them in calculator.js (#42)

It happens due to different precision between native layer and JS layer.
So, we should set expected value to calculated value in JS code.

ISSUE=none

romandev committed:
Fix a build error in calculator_bridge.h. (#45)

It is caused because we are using std::unique_ptr<> without including
<memory> header file. We should include it.

ISSUE=none

romandev committed:
Stop using C based N-API in NativeTypeTraits. (#48)

This change stops using C based N-API in NativeTypeTratis. Instead use
C++ wrapper of N-API. Also, use the NativeTypeTraits in calculator_bridge.cc.

ISSUE=#37

romandev committed:
Make the code generator generate header files by reading IDL files. (#52)

After this patch, ./bacardi build command will print generated header file by
code generator. It has the same shape with calculator_bridge.h.

ISSUE=#51

taxookim committed:
Replace "Lunch Class Authors" with "Bacardi Authors" in boilerplates. (#50)

We should replace "Lunch Class Authors" with "Bacardi Authors" in boilerplates
of each files.

ISSUE=#46

hwanseung committed:
Support build in windows platform. (#49)

After this patch, we can try to build in Windows platform as well.
This patch includes auto-test bots using AppVeyor.

ISSUE=#26

hwanseung committed:
Update README.md for windows support (#56)

Add build badge for AppVeyor and add system requirements of Windows.

ISSUE=#26

romandev committed:
Introduce JSTypeTraits to convert NativeValue to JSValue. (#57)

This change is implementing an initial codes for JSTypeTraits to convert
Native value to JS value easier. After this patch, we can use JSTypeTraits
instead of individual converting in example codes.

ISSUE=#53

romandev committed:
Should generate header/cpp file by code generator. (#59)

After this patch, ./bacardi build command will generate header and cpp file by
code generator. It's not included in actual build process but it has the same
shape with calculator_bridge.h calculator_bridge.cc.

ISSUE=#51

hwanseung committed:
Fix the nunjucks issue in windows platform. (#60)

Resolve the nunjucks issue in windows platform. Also, generated header and cpp
files will be located bacardi/build/Release/gen/examples.

ISSUE=#26

hwanseung committed:
Add header and modify boilerplates at interface_cpp.njk (#61)

Add header and modify boilerplates at interface_cpp.njk

ISSUE=none

hwanseung committed:
Should use python command in order to execute git-clang-format. (#64)

We have to use python command in order to execute git-clang-format in
windows platform explicit. But linux and mac were not needed to use
python command.

ISSUE=#26

romandev committed:
Remove examples/calculator_bridge.* files. (#62)

We can remove the examples/calculator_bridge.* files eventually. Instead use
the calculator_bridge files generated by the generator. Also, this patch fixes
much problems that occurs during trying to enable this.

ISSUE=#51

daehyunjimmy committed:
Need to change fixed code to generated by template (#66)

In interface_header.njk & interface_cpp.njk, fixed code need to be
generated by template.

ISSUE=#63

romandev committed:
Add support for static method in Interface. (#69)

Currently, we are already supporting instance methods but not static
methods. This PR is adding support for static methods in Interface.

ISSUE=#67

yjaeseok committed:
Add test framework for C++ codes (googletest) (#58)

We need a test framework for C++ codes. So I add googletest package
as submodule.

ISSUE=#47

hwanseung committed:
Fix clean command issue in appveyor (#73)

Use node-gyp.cmd instead of node-gyp for clean command in windows platforms.

ISSUE=#26

romandev committed:
Replace anonymous function with arrow function. (#74)

Stop using anonymous function for readability.

ISSUE=none

romandev committed:
Fix a bug when static method has some parameters. (#75)

In the current implementation, if a static method has some parameters,
the bridge class calls an instance method instead of a static method.
So, we should fix the bug by checking member.type and member.is_static.

ISSUE=none

romandev committed:
Add support for string type in WebIDL. (#76)

After this patch, we can use string type in bridge classes between
WebIDL and C++ implementation.

ISSUE=#68

gogag2 committed:
Need more primitive types. (#81)

In the current implementation, we only support double type as an IDLType.
So, we need more primitive types such as integer, boolean.

ISSUE=#68

yjaeseok committed:
Implement Camelcase Filter in nunjucks (#83)

We cannot use camelCase method name before
So I add Camelcase Filter in nunjucks
After this patch you can use camelcase filter in nunjucks file

e.g. (template/interface_header.njk:Line 35)

Napi::Value {{member.name | camelcase-}}(const Napi::CallbackInfo& info);

you can use camelcase filter instead of uppercase

ISSUE=#82

newklus committed:
Remove remplate/example.njk (#85)

Just removed the unnecessary example files.

ISSUE=#77

hwanseung committed:
Introduce Fragments class which is to be upper concept of interface. (#84)

Current implementation is focused to interfaces, but Fragments are needed
to implement other things.
(eg. enums, dictionaries, typedefs and implements statements)

https://heycam.github.io/webidl/#dfn-idl-fragment

ISSUE=#80

hwanseung committed:
Fix up some problem in idl.ts (#91)

This patch includes the following things:
  - Export DefinitionType enum in order to use it at outside.
  - Move to initialization position of |definitions|.
    Currently, new Array was assiged whenever iterating,
    so definitions can have only last one item

ISSUE=none

hwanseung committed:
Generate one more bridge classes when one more interfaces in idl. (#89)

Generate one more bridge classes when one more interfaces in idl.

ISSSUE=#88

daehyunjimmy committed:
Generate Bacardi files by code generator (#92)

The bacardi files should be generated by code generator. This patch is
for generating bacardi files.

ISSUE=#65

yjaeseok committed:
Implement Constructor in WebIDL (#93)

We need to implement constructor in WebIDL.
So I just implement simple constructor.

But we need to implement method overloading mechanism.

I refer to W3C Spec

ISSUE=#70

hwanseung committed:
Support snake_case in njk files. (#96)

If interfaces has CamelCase name in idl files, it have to change to
snake_case for cpp files name and Include guards. So we should support
to snake_case filter in nunjucks files.

ISSUE=#94

hwanseung committed:
Remove DefinitionType enum (#97)

Definition Type is not necessary in order to figure out type because we
can use instanceof keyword in typescript. And I think this keyword is
more clearly better than to check type attribute of Definition.

ISSUE=none

romandev committed:
Introduce reader concept and implement simple_reader.ts. (#103)

The reader is IDL files reader. The module is responsible for reading
the idl files at the proper timing. But for now, making implementation
easier, just implement simple_reader.ts. The simple implementation just
reads all IDL files without pipelining and then unifies them into one
simple data.

ISSUE=#101

corona10 committed:
Add support for long long type in WebIDL (#106)

After this patch, we can use longlong type in bridge classes between
WebIDL and C++ implementation.

ISSUE=#71

hwanseung committed:
Move generated bacardi.cc path to binding.gyp (#105)

bacardi.cc file is auto generated. actually it is related with idl files.
so present bacardi.cc path was wrote in example.gypi for gyp build.
it means bacardi user have to write bacardi.cc file path in their's gypi files.
it is better to hide auto-genreated files to user.

ISSUE=#104

Commit:
Add some badges in README.md (#102)

Add two badges for Slack link and Apache license link in README.md

ISSUE=#95

hwanseung committed:
Should take string argument in print function of Calculator class (#107)

Should take string argument in print function of Calculator class.[1]
but it was not detected because there was missed error handling for
string argument.

[1] https://github.com/lunchclass/bacardi/blob/master/examples/calculator.idl#L21

ISSUE=none

romandev committed:
Refactoring: Split idls.ts into smaller files. (#109)

Along with increasing external contribution, the complexity of the file
has increased as well. On the other hand, readability has decreased.
So, we should split the file into smaller files and clear their purpose.

The main goal of this change is spliting the idls.ts file into serveral
files and removing it.

ISSUE=#101

yjaeseok committed:
Introduce overloading feature (#108)

We need to implement overloading feature.
First I implement constructor overloading with just argument count.
Next, we should check argument type also.

ISSUE=#79

romandev committed:
Improve ./bacardi format command. (#111)

We already have ./bacardi format command to format our source code
automatically. But it works only when the change is in stage.
For example, if someone already commits some changes locally, there is
no opportunity to format the source code. So, this patch increases the
range of applying format. (HEAD~20)

Also, this patch is fixing some format errors.

ISSUE=none

romandev committed:
Add original IDL information (filename and location) to IDL definition (#112)

This is removing hard-coded value(examples/) in generateInterface() in
generator/main.ts. This is also resolving a problem that the CWD path is
differnt in Windows during generating.

ISSUE=#101

romandev committed:
Add test/test_interface.* for testing. (#113)

This patch is adding a new test IDL interface to test/ directory.
After this patch, we can build a multiple IDL files and use them.

ISSUE=#99,#100
TBR=@hwanseung,@yjaeseok

romandev committed:
Introduce jest for testing. (#114)

We are already using Mocha as testing framework but for some reasons
such as readability and convenience, we decided to use jest instead of
Mocha.

After this patch, ./bacardi test will run both of Mocha and jest.
Once we write test codes enough in jest, we will deprecate Mocha.

ISSUE=#99,#100
TBR=@hwanseung,@yjaeseok

romandev committed:
Object instantiation is wrong. (#115)

In the bridge classes, we should keep the instance of impl class.

ISSUE=none

romandev committed:
Add some tests for constructor in IDL interface. (#116)

Rewrite examples/ tests using jest.

ISSUE=#99,#100

romandev committed:
Write a test for static method in IDL interface. (#117)

This patch also changes getCalledConstructorInfo() to static method and
rename it with getLastCallInfo() and then use it use generally.

ISSUE=#99,#100

romandev committed:
Write more tests in interface_static_method.test.ts (#118)

Need more tests.

ISSUE=#99,#100

romandev committed:
Write a test for Basic Types. (#119)

Write a test for Basic Types such as boolean, short, double, and string.
Not include long type testing in this patch because it has some problem.
We should fix the issue in follow-up patch.

ISSUE=#99,#100

hwanseung committed:
Implement IDL enum feature (#110)

Implement IDL enum feature. Some part of codes should be generated
automatically at another files but this patch doesn't do that yet.
So, it will be finished in follow-up patch.

ISSUE=#80

romandev committed:
Enable --verbose option in test by default. (#121)

It's better to inform details.

ISSUE=#99,#100

hwanseung committed:
auto genereate

hwanseung committed:
Write a test for enum type using jest. (#122)

Write a test for enum type using jest.

ISSUE=#99,#100

corona10 committed:
Generate JSTypeTraits by macro functions (#123)

Generate JSTypeTraits by macro functions

In the future, we can auto-generate js_type_traits during bacardi build
time by using some information of types with template.

e.g
types = {
  'Number': {
    int32_t,
    int64_t,
    float,
    long
    .
    .
  },
  'String': {
     char*,
     std::string,
     .
   }
  
}
And this is the just pseudo code for concept.

if type in types['Number']:
   generate("JS_TYPE_TRAITS_NUMBER("+type+")");

if type in types['String']:
   generate("JS_TYPE_TRAITS_STRING("+type+")");

ISSUE=#120

romandev committed:
Add support for Electron example (#128)

This patch includes the following things.
  - Add electron and electron-rebuild
  - Patch file to disable using internal N-API in node-addon-api.
  - Electron example
  - Electron build scripts.

After this patch, we can build the example and then run as follows:
  $ ./bacardi build_electron
  $ ./bacardi electron

ISSUE=#129

hwanseung committed:
Replace loop.index - 1 with loop.index0 (#126)

Usually programmers used to count from zero when counting index.
But loop.index count from 1. so we have to minus 1 to make 0.
Actually Nunjunks provide loop.index0 which count from 0.

https://mozilla.github.io/nunjucks/templating.html#for

ISSUE=none

romandev commented on template/native_enum_type_traits.njk:
```diff
@@ -0,0 +1,73 @@
+
+{% for enum in enums %}
+template <>
+struct NativeTypeTraits<IDL{{enum.name}}>
```
IMHO, the enum type can be just unique enum type(because it's just string) and there is no reason to be NativeTypeTraits.
This generating code will generate C++ code per one enum type. It generates too much codes and can make build time increased.

We should check whether the type is enum or not in generator first.
And then, we make other logics fixed and just generate enumuration strings by generator.
For examples,
```
// This is not actual code, so, I ignores convention and grammar in my convenience.
{% if type == 'enum' %}
const char* validEnums = [
{% for each enum strings %}
    {%e enum_string %}
{% endfor %}
];
if (!IsValidEnum(validEnums, inputEnum)) return;
{% else %}
// Previsous codes
{% endif %}
```

romandev commented on template/native_enum_type_traits.njk:
```diff
@@ -0,0 +1,73 @@
+    return js_value.ToString().Utf8Value();
+  }
+
+  static bool IsTypeEquals(const Napi::Value& js_value) {
```
This is an another problem. IsTypeEquals() is used for method overloading check, right?
If developers defines two enums and then two method overloading for each enums, unexpected behavior might occur.
```
enum A {
  ...
};

enum B {
  ...
};

void method(A a);
void method(B b);
```
In WebIDL level, they seems to have different signature but the enum is just string in JS level. So, they can not be method-overloaded. So, it's a reason that we should not generate this file repeatedly.

yjaeseok committed:
Fix No repository, license warnings. (#131)

There is no repository and license field in package.json files.
For this reason we get some warnings when build bacardi project.
  npm WARN bacardi No repository field.
  npm WARN bacardi No license field.

So I add repository and license information in package.json file.

  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/lunchclass/bacardi"
  }

ISSUE=none

nadongguri committed:
Fix the typo in README.md (#132)

Modified the comment "build_electrona" as "build_electron".

ISSUE=none

yjaeseok committed:
Fix repository url (#133)

Previous PR #131 has some error.
repository url must include ".git".
I refer to gulp-typescript github page

ISSUE=none

romandev committed:
Add SimRank project for Electron example. (#134)

Original Repository:
  - https://github.com/roukaour/simrank

ISSUE=#129

romandev committed:
Add JS binding SimRank library. (#135)

This change is a good example of how we can easily bind Native code to
JS side if using Bacardi.

ISSUE=#129

hwanseung committed:
change

romandev commented:
@hwanseung Hi Hwanseung, did you come back? If you have a spare time, please re-upload this patch in new PR. (Open PRs were broken due to my mistakes.. :( )
